### ISSUE TITLE
Make use of enhanced filter possibilities

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#10027]
## Strongly linked to the [Filtering / searching project](https://github.com/cgeo/cgeo/projects/16):

The GC live map provides a bunch of filter options, which would allow "live filtering" without the need of using the "Show as list" workaround which is ugly and doesn't allow filtering by attributes if caches are not downloaded. Even basic members could filter for multiple cache types, premium members even for size, attributes, etc.

@cgeo/team Is someone interested in tackling this issue? We have a bunch of issues open, which are dealing with filtering in general. Of course, at the moment SAF has priority, but as soon as it is done, we could start with refactoring all c:geo filter functionalities, so they are easier to use and understand. 

It would be an important step for a redesign (#10003), probably even the biggest weak point in our current UI. However, I think we need to tackle this issue together as it's probably difficult to handle for a single person...

![grafik](https://user-images.githubusercontent.com/64581222/108070357-b889a100-7064-11eb-863e-79995adc7550.png)

**Comment by eddiemuc** [comment]
This one is very tricky. New filtering framework allows filtering by cache (as in the past) and filtering by SQL (for offline caches), but it does not allow filtering using a connector.

**Comment by eddiemuc** [comment]
Checked the code. If I understand current implementation correctly, this will be a BIG challenge on its own!

My understand is that c:geo currently uses an old gc.com page for searching which is based on Url https://www.geocaching.com/seek/nearest.aspx. This url seems to support searched for single capacities using parameters (e.g. "key" to search for keyword, "u" to search for owner), but it does not seem to have capabilities to combine multiple attributes (e.g. using both "u" and "key" seems to result in search only using "u" parameter) and it does not seem to have the capabilities found in todays gc.com website (e.g. no search for attributes seems to be possible)

The website gc.com nowadays seems to support a whole new search page based on Url https://www.geocaching.com/play/search. There, a whole different set of search parameters seems to be supported (e.g. "kw" for keyword, "types" for gc types, "att" for attributes and so on. The result page also looks much different from the one we currently used

Long story short: I fear to enable this we have to re-implement large parts of the gc.com search and start scanning new pages.

**Comment by Lineflyer** [comment]
> Long story short: I fear to enable this we have to re-implement large parts of the gc.com search and start scanning new pages.

That might be wise anyway as I guess they will discontinue some of the old search pages sooner or later. They already asked us years ago as they planned to retire the old search but kept it as we told them, that those are central part of c:geo.

**Comment by eddiemuc** [comment]
Digged deeper into the code and found that the Map Search (searching for new caches) uses even another API based on https://www.geocaching.com/api/proxy/web/search. It returns json and is thus much nicer to use of course than a webpage.

The Uri takes at least the parameter "box" to specify a viewport (of course, it is used for map viewer after all). Can't tell from code if this api supports filtering, but if it does it might be an option to also change "normal" search to this one. Then the whole thing wold become a lot more easy (and we can throw out some outdated code on the way)

Is there anyone who can explain me this API, and why it is (as of now) not used for "normal" searches (via keyword, owner etc)?
Maybe @moving-bits ?

**Comment by eddiemuc** [comment]
Third finding (sorry, I need to capture these thimgs somewhere so I am able to find them later):
the map on gc.com seems to use a version2 of above's Url:
https://www.geocaching.com/api/proxy/web/search/v2

This one at least seems to support search parameters from the "advanced" filter settings of gc.com. E.g. searching without any filters triggers something like this:
https://www.geocaching.com/api/proxy/web/search/v2?skip=0&take=500&asc=false&sort=distance&hf=0&sa=1&origin=xxx&fb=yyy

In contrast, searching for both cache types (Traditional +Mullti) PLUS Attribute wheelchair-accessible yields this Url:
https://www.geocaching.com/api/proxy/web/search/v2?skip=0&take=500&asc=false&sort=distance&ct=2%2C3%2C9%2C3773&hf=0&sa=1&att=24&origin=xxx&fb=yyy

(In above Urls I hid origin (my coordinates) and fb (username))

So would it be an option to change searches to gc.com completely to this Api? (Not just for maps)

**Comment by eddiemuc** [comment]
> > Long story short: I fear to enable this we have to re-implement large parts of the gc.com search and start scanning new pages.
> 
> That might be wise anyway as I guess they will discontinue some of the old search pages sooner or later. They already asked us years ago as they planned to retire the old search but kept it as we told them, that those are central part of c:geo.

@Lineflyer if they want us to retire the old search, then I wonder if this gives us some leverage to ask for official documentation of the https://www.geocaching.com/api/proxy/web/search/v2 interface? I would have a few questions on it e.g:
* how do they convert "multi-selection" of cache types into a value (in above Url you see the value 2%2C3..., it would be good to knopw how to calculate these ourself instead of storing all possible combninations in c:geo code)
* same for attributes. "wheelchair accessible" seems to be "24", would be good to know the values of the others (instead of try-and-write-down...)
* what options are supported for filtering in total, what options are supported for sorting (above they use "distance" which hopefully is equivalent to "get nearest caches"?
* Generally speaking: would they somehow keep this Api stable for us if we switch?

**Comment by Lineflyer** [comment]
> @Lineflyer if they want us to retire the old search, then I wonder if this gives us some leverage to ask for official documentation of the https://www.geocaching.com/api/proxy/web/search/v2 interface? I would have a few questions on it e.g:

I guess the answer will be, that this API is internal and we should rather use the official one :( 
I think we need to "reverse engineer" ourselves or simply stick to what we have for now.

**Comment by eddiemuc** [comment]
For my own reference: searching for different cache types reveals the following different parameter settings for ct:

- Tradi: ct=2%2C9%2C3773
- Multi: ct=3
- mystery: ct=8
- Tradi+Multi: ct=2%2C3%2C9%2C3773
- Multi+Mystery: ct=3%2C8

I conclude: 
* different cache types are separated with %2C (which is an encoded comma)
* Multi seems to have type 3, Mystery seems to have type 8, Trade seems to map to three DIFFERENT ids: 2, 9 and 3773

This seems to be decodable :-)

**Comment by moving-bits** [comment]
Have a look at `enumerations/CacheType` - are those the same identifiers?

**Comment by fm-sys** [comment]
> So would it be an option to change searches to gc.com completely to this Api?

I think this would even be the preferred way, as these APIs do change much less often, are faster and save mobile data. All newer pages are dynamic and call such web APIs from client side. Perhaps we should even ask GS if they want to shut down the old sites more quickly? ;-) 

In any case, an exchange about which sides we still need and which we don't would be really great and would IMO help both sides a lot. 

But attention, we need to check if these APIs work for basic members too.

**Comment by eddiemuc** [comment]
> But attention, we need to check if these APIs work for basic members too. 

The v1 of this API is already used by c:geo map viewer today. But you're right, V2 needs to be checked as well.

**Comment by fm-sys** [comment]
Do you know [Macrodroid](https://play.google.com/store/apps/details?id=com.arlosoft.macrodroid&hl=de&gl=US&referrer=utm_source%3Dgoogle%26utm_medium%3Dorganic%26utm_term%3Dmakrodroid&pcampaignid=APPU_19_Z1CaYMXSKcKVsAfH4ZuABQ)? It's not about geocaching, but IMHO it does combining different actions/if-statements rather well...

**Comment by eddiemuc** [comment]
OK, I have a problem where I would need some insights into current implementation...

I tried to find out how the API proxy works for basic members. As stated before, today's map viewer already uses this map Api. A typical example URL as requested by c:geo is the following:

https://www.geocaching.com/api/proxy/web/search?box=50.954265%2C6.913003%2C50.940489%2C6.925645&take=500&asc=true&skip=0&sort=distance&origin=50.947377%2C6.919324&app=cgeo 

This is a small area in cologne where there is at least one premium cache on it: GC34AVH. A normal cache nearby (for quick search of the area) is GC8RRv3

Now the problem: when I call this Uri while logged in as a basic member **I get no coordinates returned for premium caches**. This means that basic member can't see these caches on the c:geo map. And indeed, I tried it out and c:geo does not show premium caches on the map viewer for basic members.

Question now: is this a known problem, and is c:geo supposed to work like this for basic members? 

@Lineflyer , @moving-bits , @fm-sys  can anyone enlighten me here?

**Comment by eddiemuc** [comment]
Following screenshots shows a part of the formatted JSON output (one premium cache, one basic cache) when querying the above while logged in as a basic member. As you can see, the "postedCoordinates"-entry is missing for the premium cache:

![image](https://user-images.githubusercontent.com/6909759/118171122-3fafa080-b42b-11eb-82a1-041cceb1f294.png)

**Comment by Lineflyer** [comment]
> Now the problem: when I call this Uri while logged in as a basic member I get no coordinates returned for premium caches. This means that basic member can't see these caches on the c:geo map. And indeed, I tried it out and c:geo does not show premium caches on the map viewer for basic members.

> Question now: is this a known problem, and is c:geo supposed to work like this for basic members?

Yes, it it like that since we started using this API for the map and I think its OK.

**Comment by eddiemuc** [comment]
> Have a look at `enumerations/CacheType` - are those the same identifiers? (`wptTypeId`)

Right you are. Seems that gc.com "groups" certain cache types into one "selectable" type in the filter. In case of Traditional, the rather seldom types PROJECT_APE and GCHQ (both of which I never heard of before) are included

**Comment by fm-sys** [comment]
> This means that basic member can't see these caches on the c:geo map. And indeed, I tried it out and c:geo does not show premium caches on the map viewer for basic members.
>
> Question now: is this a known problem, and is c:geo supposed to work like this for basic members?
>
>@Lineflyer , @moving-bits , @fm-sys can anyone enlighten me here?

There is simply no way to retire the coords of PM caches as basic member at all. Not a single subpage provide this information. That's how it is supposed to work (otherwise the PM-only flag would be useless), and IMHO it is even a good thing that they don't show up, as it is really annoying if you see caches but your cannot open them... (at least on the live map)

**Comment by eddiemuc** [comment]
Ok then I am confident again :-) Thanks a lot!

Another thing I didn't know: we actually identify ourselves as c:geo when calling the GC.com API. Note the parameter "app=cgeo" in the Url above

**Comment by moving-bits** [comment]
> That's how it is supposed to work (otherwise the PM-only flag would be useless), and IMHO it is even a good thing that they don't show up, as it is really annoying if you see caches but your cannot open them... (at least on the live map)

Before we switched to that API some (2?) years ago PM-only caches were shown on the live map even for BM (with unreliable coords). It's been very annoying back then, that you saw those caches, but were not able to open them, but got an error toast instead.

**Comment by moving-bits** [comment]
> Another thing I didn't know: we actually identify ourselves as c:geo when calling the GC.com API. Note the parameter "app=cgeo" in the Url above

That's intentionally, to give the receiving ends (gc, Mapsforge etc.) a chance to identify us and measure our usage.

**Comment by eddiemuc** [comment]
Solved with PR #10622 . Be aware of limitations as mentioned in #10633 and #10621. Search limit is no longer just 20.
